server/src/main/java/org/opensearch/common/util/concurrent/ThreadContext.java

```diff
@@ -710,6 +728,8 @@ private ThreadContextStruct copyHeaders(Iterable<Map.Entry<String, String>> head
         }
 
         private void writeTo(StreamOutput out, Map<String, String> defaultHeaders) throws IOException {
+            TracerManager.getTracerHeaderInjector().accept(this.requestHeaders, this.transientHeaders);
```

Comment:
1. this should be behind feature flag/or should be no-op when tracing is disabled.
2. It should not add tracer context into request headers blindly. If distributed tracing is not required for any span/operation (which can be made configurable and should be disabled as a default behavior), then this should again be no-op. 
3. If `this.requestHeaders` is empty, it would be an instance of `Collections.emptyMap()` and if there is trace context present, you will be adding entries into `Collections.emptyMap()`, which would result in error the it is handled - 
```java
        if (carrier != null) {
            carrier.put(key, value);
        }
```
I would suggest you to move this line to https://github.com/opensearch-project/OpenSearch/pull/7648/files#diff-88483cd727c5ab23cd396ed2230b494eba26a1b76d37173eff131b4a11aca8e9L720 and replace `this.requestHeaders` with `requestHeaders`

Reply:
Good point. But actually we do not add the headers blindly. We have a check if the transient headers have our key (`CURRENT_SPAN`) before injecting in the headers. And that key will not be added in the transient headers if tracing is disabled. Also while injecting span in headers if the tracing is enabled, we find the last valid span (i.e not NoopSpan) and inject it only.

Reply:
I shouldn't have used 'blindly' and I get the idea of checking the span context in thread context, I was concerned about the case where we don't need distributed tracing and just need local tracing for some operations, still we will be injecting context information into the request headers if its part of some active span.

Reply:
Why we don't inject the span to request headers (using `threadContextStruct.putHeader`) in `ThreadContext::stash` / `ThreadContextBasedTracerContextStorage::put` but defer that to `writeTo` method?

Reply:
Request headers are immutable like transient headers. But they are map of <String, String> so we cannot update its value using the workaround(SpanReference) we use for transient headers. Hence we are deferring it to `writeTo` to avoid multiple updates

Reply:
Resolving as this has been addressed by Reta's commit